ember-server/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala

```diff
@@ -59,30 +59,33 @@ private[server] object ServerHelpers {
       idleTimeout: Duration = 60.seconds,
       additionalSocketOptions: List[SocketOptionMapping[_]] = List.empty,
       logger: Logger[F]
-  )(implicit F: Concurrent[F], C: Clock[F]): Stream[F, Nothing] = {
+  )(implicit F: Concurrent[F], T: Timer[F]): Stream[F, Nothing] = {
     def socketReadRequest(
         socket: Socket[F],
         requestHeaderReceiveTimeout: Duration,
-        receiveBufferSize: Int,
-        isReused: Boolean
+        receiveBufferSize: Int
     ): F[Request[F]] = {
-      val (initial, readDuration) = (requestHeaderReceiveTimeout, idleTimeout, isReused) match {
-        case (fin: FiniteDuration, idle: FiniteDuration, true) => (true, idle + fin)
-        case (fin: FiniteDuration, _, false) => (true, fin)
-        case _ => (false, Duration.Zero)
+      val idle: Option[FiniteDuration] = idleTimeout match {
+        case idle: FiniteDuration => Some(idle)
+        case _ => None
       }
-
-      SignallingRef[F, Boolean](initial).flatMap { timeoutSignal =>
-        C.realTime(MILLISECONDS)
-          .flatMap(now =>
-            Parser.Request
-              .parser(maxHeaderSize)(
-                readWithTimeout[F](socket, now, readDuration, timeoutSignal.get, receiveBufferSize)
-              )
-              .flatMap { req =>
-                timeoutSignal.set(false).as(req)
-              })
+      def errorIfEmpty(
+          socket: Socket[F],
+          maxBytes: Int,
+          timeout: Option[FiniteDuration]): Stream[F, Byte] =
+        Stream.eval(socket.read(maxBytes, timeout)).flatMap {
```

Comment:
Out of curiosity, what happens when the client sends two requests in a row such that they are read in as part of the same chunk?

Reply:
Presently, failure. We can easily get this by moving this stream be passed in rather than created here.

Reply:
This has been added in https://github.com/http4s/http4s/pull/4301/commits/5599b528ac3cda381d6506e108d4a6a4bd92a245